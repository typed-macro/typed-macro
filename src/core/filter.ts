import { posix, win32 } from 'path'
import pm from 'picomatch'

export type FilterPattern = RegExp | string | Array<RegExp | string>

export type FilterOptions = {
  /**
   * Regexp or glob patterns to include matches, applied after
   * {@link FilterOptions.exclude}.
   *
   * The default value is `/\.[jt]sx?$/`.
   *
   * Once this option is defined, the default one will be overwritten.
   */
  include?: FilterPattern
  /**
   * Regexp or glob patterns to exclude matches, applied before
   * {@link FilterOptions.include}.
   *
   * The default value is `/node_modules/`.
   *
   * Once this option is defined, the default one will be overwritten.
   */
  exclude?: FilterPattern
}

export function createFilter({
  include = /\.[jt]sx?$/,
  exclude = /node_modules/,
}: FilterOptions) {
  const getMatcher = (pattern: string | RegExp) => {
    if (pattern instanceof RegExp) return pattern
    const fn = pm(pattern, { dot: true })
    return { test: (id: string) => fn(id) }
  }
  const ensureArray = (v: unknown) => (v ? (Array.isArray(v) ? v : [v]) : [])

  const includeMatchers = ensureArray(include).map(getMatcher)
  const hasIncludeRule = includeMatchers.length > 0
  const excludeMatchers = ensureArray(exclude).map(getMatcher)

  return (id: string) => {
    const pathId = id.split(win32.sep).join(posix.sep)
    for (let i = 0; i < excludeMatchers.length; ++i) {
      if (excludeMatchers[i].test(pathId)) return false
    }

    for (let i = 0; i < includeMatchers.length; ++i) {
      if (includeMatchers[i].test(pathId)) return true
    }

    return !hasIncludeRule
  }
}
