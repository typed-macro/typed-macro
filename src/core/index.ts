export { Runtime, RuntimeOptions } from './runtime'
export {
  createTransformer,
  Transformer,
  TransformerOptions,
} from './transformer'
export {
  Macro,
  MacroContext,
  MacroHelper,
  MacroHandler,
  Exportable,
  NamespacedExportable,
  NormalizedExports,
  normalizeExports,
} from './types'
