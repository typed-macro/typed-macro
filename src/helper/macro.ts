import { dirname, resolve, sep } from 'path'
import { existsSync } from 'fs'
import { Node, NodePath } from '@babel/traverse'
import template from '@babel/template'
import { File, ImportDeclaration, Program } from '@babel/types'
import {
  ImportOption,
  matchImportStmt,
  renderImportStmt,
} from '@/helper/import'
import { findProgramPath } from '@/common'

export type Helper = {
  /**
   * Get the directory of the root project or the nearest project.
   * @param which 'root' or 'leaf'
   */
  projectDir: (which: 'root' | 'leaf') => string

  /**
   * Normalize path pattern so can resolve file paths as import paths
   * @param path relative path or path pattern
   * @param root absolute file path of the project dir, default to projectDir('leaf')
   * @param importer absolute file path of the importer, default to current module
   *
   * e.g. search files by glob pattern and then import them
   *
   * ```typescript
   * // /src/example/a.ts
   * // /src/example/b.ts
   * // in /src/another/index.ts
   * const pattern = '../example/*.ts'
   * const { normalized, base, resolveImportPath } = normalizePathPattern(pattern)
   * const importPaths = glob.sync(normalized, { cwd: base }).map(resolveImportPath)
   *
   * // importPaths = ['../example/a.ts', '../example/b.ts']
   * ```
   */
  normalizePathPattern: (
    pattern: string,
    root?: string,
    importer?: string
  ) => {
    /**
     * Normalized path or path pattern,
     */
    normalized: string
    /**
     * Base path for the normalized path or path pattern.
     */
    base: string
    /**
     * Get import path from file path got by normalized path and base,
     * almost the opposite operation of normalization
     * @param path file path
     */
    resolveImportPath: (path: string) => string
  }

  /**
   * Prepend import statements to program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   */
  prependImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration>

  /**
   * Append import statements to the last import statement of the program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   */
  appendImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration>

  /**
   * Find an import statement that has been in the target program already.
   * Note that findImported() will not trim extension for moduleName.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  findImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Check if an import statement has been in the target program already.
   * Note that hasImported() will not trim extension for moduleName.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  hasImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => boolean

  /**
   * Prepend any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  prependToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Append any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  appendToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Get current program or the program of provided node.
   * @param node any node path
   */
  getProgram: (node?: NodePath) => NodePath<Program>

  /**
   * Force to recollect imported macros in the next loop.
   * Usually used after importing new macros to program during expanding a macro.
   *
   * Note:
   *
   *  Only collected imported macros can be applied, because plugin only
   *  searches call expressions for collected macros. Also, external macros
   *  cannot be collected because plugin is unaware of external macros.
   */
  forceRecollectMacros: () => void
}

const projectDirCache = {
  root: '',
  leaf: '',
}

const projectDir: Helper['projectDir'] = (which) => {
  if (projectDirCache[which]) return projectDirCache[which]

  const paths = (require.main?.path || process.cwd()).split(sep)

  if (which === 'root') {
    for (let i = 0; i < paths.length; i++) {
      const path = paths.slice(0, i + 1).join(sep)
      if (existsSync([path, 'node_modules'].join(sep)))
        return (projectDirCache['root'] = path)
    }
  } else {
    for (let i = paths.length; i > 0; i--) {
      const path = paths.slice(0, i).join(sep)
      if (existsSync([path, 'node_modules'].join(sep)))
        return (projectDirCache['leaf'] = path)
    }
  }

  throw new Error('can not find project root')
}

export function getHelper(
  filepath: string,
  ast: File
): Omit<Helper, 'forceRecollectMacros'> {
  const thisProgram = findProgramPath(ast)

  const normalizePathPattern: Helper['normalizePathPattern'] = (
    pattern,
    root = projectDir('leaf'),
    importer = filepath
  ) => {
    if (!pattern.startsWith('.') && !pattern.startsWith('/')) {
      throw new Error(
        `pattern must start with "." or "/" (relative to project root)`
      )
    }
    let base: string
    let parentDepth = 0
    const isAbsolute = pattern.startsWith('/')
    if (isAbsolute) {
      base = resolve(root)
      pattern = pattern.slice(1)
    } else {
      base = dirname(importer)
      while (pattern.startsWith('../')) {
        pattern = pattern.slice(3)
        base = resolve(base, '../')
        parentDepth++
      }
      if (pattern.startsWith('./')) {
        pattern = pattern.slice(2)
      }
    }
    return {
      normalized: pattern,
      base,
      resolveImportPath: isAbsolute
        ? (f) => `/${f}`
        : parentDepth
        ? (f) => `${'../'.repeat(parentDepth)}${f}`
        : (f) => `./${f}`,
    }
  }

  const findImported: Helper['findImported'] = (
    imp,
    loose = true,
    program = thisProgram
  ) => {
    // An import declaration can only be used in top-level.
    for (const path of program.get('body') as NodePath[]) {
      if (!path.isImportDeclaration()) continue
      if (matchImportStmt(imp, path.node, loose)) return path
    }
  }

  const hasImported: Helper['hasImported'] = (
    imp,
    loose = true,
    program = thisProgram
  ) => {
    return findImported(imp, loose, program) !== undefined
  }

  // returns
  // - array: generated import statements
  // - path: all import statements are duplicated, returns the node path of the last one
  function normalizeImports(
    imports: ImportOption | ImportOption[],
    program: NodePath<Program>
  ) {
    if (!Array.isArray(imports)) imports = [imports]
    // remove duplicated
    const toBeImported = imports.filter(
      (imp) => !hasImported(imp, false, program)
    )
    // all import statements are duplicated, returns the node path of the last one
    if (!toBeImported.length)
      return findImported(imports[imports.length - 1], false, program)!
    return template.statements.ast(
      Array.from(
        new Set(toBeImported.map((imp) => renderImportStmt(imp)))
      ).join('; ')
    ) as ImportDeclaration[]
  }

  const prependImports: Helper['prependImports'] = (
    imports,
    program = thisProgram
  ) => {
    const toBeImported = normalizeImports(imports, program)
    if (!Array.isArray(toBeImported)) return toBeImported
    const firstImport = (program.get('body') as NodePath[]).filter((p) =>
      p.isImportDeclaration()
    )[0]
    return (
      firstImport
        ? firstImport.insertBefore(toBeImported)
        : program.unshiftContainer('body', toBeImported)
    ).pop() as NodePath<ImportDeclaration>
  }

  const appendImports: Helper['appendImports'] = (
    imports,
    program = thisProgram
  ) => {
    const toBeImported = normalizeImports(imports, program)
    if (!Array.isArray(toBeImported)) return toBeImported
    const lastImport = (program.get('body') as NodePath[])
      .filter((p) => p.isImportDeclaration())
      .pop()
    return (
      lastImport
        ? lastImport.insertAfter(toBeImported)
        : program.unshiftContainer('body', toBeImported)
    ).pop() as NodePath<ImportDeclaration>
  }

  const prependToBody: Helper['prependToBody'] = (
    nodes,
    program = thisProgram
  ) => {
    return program.unshiftContainer('body', nodes).pop() as NodePath
  }

  const appendToBody: Helper['appendToBody'] = (
    nodes,
    program = thisProgram
  ) => {
    return program.pushContainer('body', nodes).pop() as NodePath
  }

  const getProgram: Helper['getProgram'] = (node) => {
    if (!node) return thisProgram
    return node.findParent((p) => p.isProgram()) as NodePath<Program>
  }

  return {
    projectDir,
    normalizePathPattern,
    findImported,
    hasImported,
    prependImports,
    appendImports,
    prependToBody,
    appendToBody,
    getProgram,
  }
}
