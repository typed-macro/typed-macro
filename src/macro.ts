import { Macro } from '@/runtime/types'

export type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  types: string[]
}

export type MacroWithMeta = Macro & {
  meta: MacroMeta
}

export function renderMacroType({ name, meta }: MacroWithMeta) {
  return [
    meta.types.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
