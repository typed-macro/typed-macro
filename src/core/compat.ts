/*
 * This is a compatibility control mechanism based on version number,
 * to ensure the compatibility between macro and runtime.
 */

import { Macro } from '@/core/macro'
import { WithVersion } from '@/common'

/*
 * The current Macro version.
 *
 * It represents whether the call-context passed by the Runtime to the Macro is compatible.
 * This involves not only the structure of the call-context itself,
 * but also the structure of the AST between different versions of Babel,
 * for AST is generated in Runtime but transformed in Macros.
 *
 * @deps {@link RawMacroCall}, {@link Babel}
 */
export const MACRO_VERSION = 0

export function versionedMacro(v: Macro): Macro {
  ;(v as WithVersion<Macro>).$__version = MACRO_VERSION
  return v
}

export function isMacroCompatible(v: Macro) {
  return (v as WithVersion<Macro>).$__version === MACRO_VERSION
}
