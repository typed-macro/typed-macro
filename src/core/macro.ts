import { CallExpression } from '@babel/types'
import { NodePath } from '@babel/traverse'
import { Babel, PathHelper, StateHelper, TransformHelper } from '@/core/helper'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /*
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath
  /*
   * Whether in ssr mode.
   */
  ssr: boolean
}

export type MacroHelper = TransformHelper & PathHelper & StateHelper

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

export type Macro = {
  name: string
  apply: MacroHandler
}

export type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  types: string[]
}

type InternalMacro = Macro & {
  __types: string
}

export function macro(
  name: string,
  meta: MacroMeta,
  handler: MacroHandler
): Macro {
  return {
    name,
    apply: handler,
    __types: renderMetaType(name, meta),
  } as InternalMacro
}

export function isMacro(o: unknown): o is InternalMacro {
  return !!(o as InternalMacro).__types
}

export function renderMetaType(name: string, meta: MacroMeta) {
  return [
    meta.types.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
