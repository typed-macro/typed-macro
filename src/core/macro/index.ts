import { CallExpression, File, Program } from '@babel/types'
import { NodePath } from '@babel/traverse'

import { CURRENT_CALL_VERSION, Versioned } from '@/core/version'
import { State } from '../helper/state'
import { HELPER, MacroHelper } from './helper'
import { Babel, BABEL_TOOLS } from './babel'
import { ImportedMacro } from '@/core/helper/traverse'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /**
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath<CallExpression>
  /**
   * The AST being traversed.
   */
  ast: File
  /**
   * Whether in ssr mode.
   */
  ssr: boolean
  /**
   * Traversal state.
   */
  state: {
    /**
     * In order to recursively expand all macros in a file,
     * the transformer will traverse the AST many times until all macros
     * are expanded.
     *
     * - {@link traversal} is the state shared during one traversal, and
     * - {@link transform} is the state shared during the whole transformation.
     */
    traversal: State
    /**
     * @see traversal
     */
    transform: State
  }
}

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

type RawMacroCall = Versioned<{
  filepath: string
  code: string
  path: NodePath<CallExpression>
  ast: File
  program: NodePath<Program>
  ssr: boolean
  traversalState: State
  transformState: State
  importedMacros: ImportedMacro[]
}>

export type Macro = (call: RawMacroCall) => void

export type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  types: string[]
}

type InternalMacro = Macro & {
  __types: string
}

export function macro(
  name: string,
  meta: MacroMeta,
  handler: MacroHandler
): Macro {
  const normalizer = getNormalizer(handler)
  const m = {
    [name](call: RawMacroCall) {
      const { ctx, babel, helper } = normalizer(call)
      handler(ctx, babel as Babel, helper as MacroHelper)
    },
  }[name]
  ;(m as InternalMacro).__types = renderMetaType(name, meta)
  return m
}

type MacroCallNormalizer = (raw: RawMacroCall) => {
  ctx: MacroContext
  babel?: Readonly<Babel>
  helper?: Readonly<MacroHelper>
}

function ensureCompatible(raw: RawMacroCall) {
  if (raw.__version !== CURRENT_CALL_VERSION)
    throw new Error('the macro is incompatible with the runtime')
}

function normalizeContext(raw: RawMacroCall): MacroContext {
  const { filepath, ssr, ast, code, path, transformState, traversalState } =
    raw.value
  return {
    filepath,
    code,
    ast,
    path,
    args: path.node.arguments,
    state: {
      transform: transformState,
      traversal: traversalState,
    },
    ssr,
  }
}

export function getNormalizer(handler: MacroHandler): MacroCallNormalizer {
  switch (handler.length) {
    case 0:
      throw new Error('not a valid macro handler')
    case 1:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        return {
          ctx: normalizeContext(raw),
        }
      }

    case 2:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        return {
          ctx: normalizeContext(raw),
          babel: BABEL_TOOLS,
        }
      }

    default:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        const ctx = normalizeContext(raw)
        const { path, filepath, program, importedMacros } = raw.value
        HELPER.props
          .setImportedMacros(importedMacros)
          .setPath(path)
          .setFilepath(filepath)
          .setProgram(program)
        return {
          ctx,
          babel: BABEL_TOOLS,
          helper: HELPER.instance,
        }
      }
  }
}

export function isMacro(o: unknown): o is InternalMacro {
  return (o as InternalMacro)?.__types !== undefined ?? false
}

export function renderMetaType(name: string, meta: MacroMeta) {
  return [
    meta.types.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
