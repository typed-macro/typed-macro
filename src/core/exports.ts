import { isMacro, Macro } from '@/core/macro'
import { isMacroCompatible } from '@/core/compat'

export type NamespacedMacros = { [namespace: string]: Macro[] }

export type NamespacedModules = { [namespace: string]: string }

export type NamespacedTypes = {
  [namespace: string]: {
    moduleScope: string[]
    macroScope: string[]
  }
}

export type Exportable = (
  | {
      /**
       * as macros
       */
      macros: Macro[]
    }
  | {
      /**
       * as module, code will be used as source code of the module
       */
      code: string
    }
) & {
  /**
   * Type definitions, will be written to d.ts.
   *
   * e.g.
   * ```typescript
   * { exports: { '@macros': { customTypes: `type A = string` } } }
   * ```
   * will generate
   * ```typescript
   * declare module '@macros' {
   *   type A = string
   * }
   * ```
   */
  customTypes?: string
}

export type NamespacedExportable = { [namespace: string]: Exportable }

export type NormalizedExports = {
  macros: NamespacedMacros
  modules: NamespacedModules
  types: NamespacedTypes
}

export function normalizeExports(
  exports: NamespacedExportable
): NormalizedExports {
  const macros: NamespacedMacros = Object.create(null)
  const modules: NamespacedModules = Object.create(null)
  const types: NamespacedTypes = Object.create(null)
  Object.keys(exports).forEach((ns) => {
    const item = exports[ns]
    types[ns] = {
      moduleScope: item.customTypes ? [item.customTypes] : [],
      macroScope: [],
    }
    if ('code' in item) {
      modules[ns] = item.code
    } else {
      macros[ns] = item.macros
      types[ns].macroScope = item.macros.map((m) => {
        if (!isMacro(m)) throw new Error(`'${m.name}' is not a macro`)
        return m.__types
      })
    }
  })

  validateMacros(macros)

  return {
    macros,
    modules,
    types,
  }
}

// ensure compatible and no name conflict
export function validateMacros(macros: NamespacedMacros) {
  Object.keys(macros).forEach((ns) => {
    const mem = Object.create(null)
    macros[ns].forEach((m) => {
      if (!isMacroCompatible(m))
        throw new Error(`macro '${m.name}' in '${ns}' is incompatible`)

      if (mem[m.name])
        throw new Error(`macro '${m.name}' in '${ns}' is duplicated`)

      mem[m.name] = 1
    })
  })
}
