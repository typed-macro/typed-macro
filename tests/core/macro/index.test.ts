import { isMacro, macro, renderMetaType } from '@/core/macro'
import { NO_OP, NO_OP_HANDLER } from '../../testutils'

describe('renderMetaType()', () => {
  it('should work', () => {
    expect(
      renderMetaType('macro', {
        signatures: [
          {
            signature: `(): A`,
            comment: `hello`,
          },
        ],
        types: ['type A = string'],
      })
    ).toMatchSnapshot()
  })
})

describe('macro()', () => {
  it('should throw error if handler is invalid', () => {
    expect(() =>
      macro(
        'test',
        {
          signatures: [
            {
              signature: '_',
            },
          ],
          types: [],
        },
        // NO_OP has no parameter
        NO_OP
      )
    ).toThrow()
  })
})

describe('macro() & isMacro()', () => {
  it('should work', () => {
    expect(
      isMacro(
        macro(
          'test',
          {
            signatures: [
              {
                signature: '_',
              },
            ],
            types: [],
          },
          NO_OP_HANDLER
        )
      )
    ).toBe(true)
    expect(isMacro(undefined)).toBe(false)
    expect(isMacro(false)).toBe(false)
  })
})

describe('Macro', () => {
  it('should throw error when incompatible with the runtime', () => {
    expect(() =>
      macro(
        'test',
        {
          signatures: [
            {
              signature: '_',
            },
          ],
          types: [],
        },
        NO_OP_HANDLER
      )({ __version: -1, value: {} as any })
    ).toThrow()
  })
})
