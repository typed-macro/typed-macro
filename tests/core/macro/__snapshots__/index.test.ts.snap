// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderMetaType() should work 1`] = `
"type A = string
  /* hello */
  export function macro(): A"
`;
